nuklear_console_file.h

```diff
@@ -0,0 +1,417 @@
+#define NK_CONSOLE_FILE_PATH_MAX 1024
+#endif
+#endif  // NK_CONSOLE_FILE_PATH_MAX
+
+/**
+ * Custom data for the file widget.
+ */
+typedef struct nk_console_file_data {
+    nk_console_button_data button; /** Inherited from Button */
+    char* file_path_buffer; /** The string buffer for the chosen file path. */
+    int file_path_buffer_size; /** The size of the buffer. */
+    char directory[NK_CONSOLE_FILE_PATH_MAX]; /** When selecting a file, this is the current directory. */
+    void* file_user_data; /** Custom user data for the file system. */
+} nk_console_file_data;
+
+/**
+ * Creates a file widget that allows the user to select a file.
+ *
+ * @param parent The parent widget.
+ * @param label The label for the file widget. For example: "Select a file".
+ * @param file_path_buffer The buffer to store the file path.
+ * @param file_path_buffer_size The size of the buffer.
+ *
+ * @return The new file widget.
+ */
+NK_API nk_console* nk_console_file(nk_console* parent, const char* label, char* file_path_buffer, int file_path_buffer_size);
+
+/**
+ * Render callback to display the file widget.
+ */
+NK_API struct nk_rect nk_console_file_render(nk_console* widget);
+
+/**
+ * Sets any custom user data specifically for the file widget.
+ *
+ * This can be helpful for storing additional file system data for the file widget.
+ *
+ * @param file The file widget.
+ * @param user_data The custom user data.
+ */
+NK_API void nk_console_file_set_file_user_data(nk_console* file, void* user_data);
+
+/**
+ * Gets the custom user data specifically for the file widget.
+ *
+ * @param file The file widget.
+ *
+ * @return The custom user data.
+ */
+NK_API void* nk_console_file_get_file_user_data(nk_console* file);
+
+/**
+ * Add a individual file or directory to the given file widget as a child.
+ *
+ * This should be called from the file system callbacks. See `nuklear_console_file_system.h` for examples.
+ *
+ * @param parent The file widget.
+ * @param path The path to the file or directory.
+ * @param is_directory True if the path is a directory. False otherwise.
+ *
+ * @see nk_console_file_destroy_tinydir()
+ */
+NK_API void nk_console_file_add_entry(nk_console* parent, const char* path, nk_bool is_directory);
+
+/**
+ * Refreshes the file widget to display the contents of the current directory.
+ *
+ * @param widget The file widget to refresh.
+ *
+ * @see nk_console_file_data::directory
+ */
+NK_API void nk_console_file_refresh(nk_console* widget);
+
+#if defined(__cplusplus)
+}
+#endif
+
+#endif  // NK_CONSOLE_FILE_H__
+
+#if defined(NK_CONSOLE_IMPLEMENTATION) && !defined(NK_CONSOLE_HEADER_ONLY)
+#ifndef NK_CONSOLE_FILE_IMPLEMENTATION_ONCE
+#define NK_CONSOLE_FILE_IMPLEMENTATION_ONCE
+
+#if defined(__cplusplus)
+extern "C" {
+#endif
+
+/**
+ * Gets the base name of a file path.
+ */
+static const char* nk_console_file_basename(const char* path) {
+    if (path == NULL) {
+        return NULL;
+    }
+
+    // TODO: Ensure UTF-8 compatibility.
```

Comment:
This looks fine for UTF-8 because all ASCII character encodings are never used in the extended code bytes (second, third, and fourth code bytes always have their highest bits set)